Enhanced JetStream documentation (#14150)
kyma-project/kyma: docs/03-tutorials/00-eventing/evnt-03-introduction-to-jetstream.md

---
title: Introduction to JetStream
---

JetStream is the 'built-in distributed persistence system' of NATS. You can read more about it [here](../../05-technical-reference/00-architecture/evnt-01-architecture.md#jetstream).

## Streams and Consumers

> A `Stream` is a 'message store' for the events published. In Kyma, we use only one stream for all the events. You can configure the retention and delivery policies for the stream, based on the use-case.

Author: I think it's just a matter of emphasis. The first statement emphasises events and the second one on published. Both are right in my opinion. I like the second one (your recommendation) better for this case.

> A `Consumer` is a 'view' into the stream. A kyma subscription creates one consumer for each filter specified. In Kyma we use push based consumers.

Author: You mean ".." vs '..'? I just like single quotes in this case as its just a light emphasis and I'm not quoting something or someone. But lets see what the technical writers say :)

> A `Consumer` is a 'view' into the stream. A kyma subscription creates one consumer for each filter specified. In Kyma we use push based consumers.

Reviewer: Yeah lets let them decide :)

## Steps to verify at least once delivery with JetStream backend

> 1. Create a [Function](../../02-get-started/04-trigger-workload-with-event.md#create-a-function), [Subscription](../../02-get-started/04-trigger-workload-with-event.md#create-a-subscription) and [trigger the workload with an event](../../02-get-started/04-trigger-workload-with-event.md#trigger-the-workload-with-an-event).

Reviewer: I think we need a little intro section what this tutorial is about. Basically explaining the purpose and what the reader will learn here.
Something like "The following tutorial will showcase the at least once delivery guarantee of JetStream by pausing the Subscription sink while sending an event"

    1. Port forward the Kyma Eventing NATS Service to localhost. Use port `4222`. Run:
       ```bash
       kubectl -n kyma-system port-forward svc/eventing-nats 4222
       ```
    2. Use the [nats CLI](https://github.com/nats-io/natscli) to list the streams:
       ```bash
       nats stream ls
       ```

       Notice that the stream contains undelivered messages:
       ```bash
        ╭────────────────────────────────────────────────────────────────────────────╮
        │                                  Streams                                   │
        ├──────┬─────────────┬─────────────────────┬──────────┬───────┬──────────────┤
        │ Name │ Description │ Created             │ Messages │ Size  │ Last Message │
        ├──────┼─────────────┼─────────────────────┼──────────┼───────┼──────────────┤
        │ sap  │             │ 2022-04-26 00:00:00 │ 1        │ 318 B │ 5.80s        │
        ╰──────┴─────────────┴─────────────────────┴──────────┴───────┴──────────────╯
       ```

> 5. Recreate your [Function](../../02-get-started/04-trigger-workload-with-event.md#create-a-function). The stream no longer contains any undelivered messages and the event is delivered. Follow the [tutorial](../../02-get-started/04-trigger-workload-with-event.md#verify-the-event-delivery) to verify the event delivery.

Reviewer: We could summarise what happened here: "even though the sink was not reachable, JetStream did persist the event and redelivered it then the sink was available again, therefore the at-least-once delivery promise was met".

Not sure if we need this in addition to the intro section I proposed in another comment.

Author: I partially used this sentence and wrote it as an introduction to the section.

---
title: Introduction to JetStream
---

JetStream is the 'built-in distributed persistence system' of NATS. You can read more about it [here](../../05-technical-reference/00-architecture/evnt-01-architecture.md#jetstream).

## Streams and Consumers

> A `Stream` is a 'message store' for the published events. In Kyma, we use only one stream for all the events. You can configure the retention and delivery policies for the stream, based on the use-case.

Reviewer 2: ```suggestion
A Stream specifies how messages are stored for the published events. In Kyma, we use only one stream for all the events. You can configure the retention and delivery policies for the stream, depending on the use case.
```

Author: It doesn't specify how it stores. It basically is the store.

Reviewer 2: Ok, how about now?

Author: This is better!

> A `Consumer` is a 'view' into the stream. A kyma subscription creates one consumer for each filter specified. In Kyma we use push based consumers.

Reviewer 2: ```suggestion
A Consumer gives you an insight into the stream. Kyma Subscription creates one consumer for each filter specified. In Kyma we use push based consumers.
```

Author: Here as well: It is the view into the stream. That's why I used the single quotes.

Reviewer 2: So it's a view into the stream, but does not give you an insight? Can you clarify how it works then?

Author: It's just a technical wording in my opinion. Its not an insight (I would use this term more for metrics) but this is more like `looking into the message store` if some event is meant for you

> A Consumer reads or consumes the messages from the stream. Kyma Subscription creates one consumer for each filter specified. In Kyma we use push based consumers.

Reviewer 2: ```suggestion
A Consumer reads or consumes the messages from the stream. Kyma Subscription creates one consumer for each filter specified. In Kyma we use push-based consumers.
```

## Steps to verify at least once delivery with JetStream backend

> In this tutorial, we will verify how JetStream persists events even when the sink is not reachable, and redelivers the event when the sink is available again.

Reviewer 2: ```suggestion
This tutorial shows how JetStream persists events, even when the sink is not reachable, and redelivers the event when the sink is available again.
```
